code/datums/status_effects/status_effect.dm

```diff
@@ -277,3 +277,24 @@
 /datum/status_effect/grouped/before_remove(source)
 	sources -= source
 	return !length(sources)
+
+//These effects reapplies their on_apply() effect when refreshed while stacks < max_stacks
+/datum/status_effect/limited_buff
+	id = "limited_buff"
+	duration = -1
+	status_type = STATUS_EFFECT_REFRESH
+	///How many stacks we currently have
+	var/stacks = 1
+	///How many stacks we can have maximum
+	var/max_stacks = 3
+
+/datum/status_effect/limited_buff/refresh(effect, ...)
+	. = ..()
```

Comment:
since limited buffs don't have duration, you can actually probably cut this since it doesn't do anything